Extension/src/Utility/Text/scanner.ts

```diff
@@ -0,0 +1,1018 @@
+
+        // end when the first column is triple backtick
+        const code = this.scanUntil((ch, chNext, chNextNext) => this.#column === 0 && (ch === CharacterCodes.backtick && chNext === CharacterCodes.backtick && chNextNext === CharacterCodes.backtick));
+
+        this.text = fenceStart + code;
+
+        // code is just the inside contents
+        this.stringValue = code.substring(0, code.length - 3);
+
+        return this.kind = Kind.CodeFence;
+    }
+
+    private unescapeString(text: string) {
+        let result = '';
+        let start = 0;
+        let pos = 0;
+        const end = text.length;
+
+        while (pos < end) {
+            let ch = text.charCodeAt(pos);
+            if (ch !== CharacterCodes.backslash) {
+                pos++;
+                continue;
+            }
+
+            result += text.substring(start, pos);
+            pos++;
+            ch = text.charCodeAt(pos);
+
+            switch (ch) {
+                case CharacterCodes.r:
+                    result += '\r';
+                    break;
+                case CharacterCodes.n:
+                    result += '\n';
+                    break;
+                case CharacterCodes.t:
+                    result += '\t';
+                    break;
+                case CharacterCodes.singleQuote:
+                    result += '\'';
+                    break;
+                case CharacterCodes.doubleQuote:
+                    result += '"';
+                    break;
+                case CharacterCodes.backslash:
+                    result += '\\';
+                    break;
+                case CharacterCodes.backtick:
+                    result += '`';
+                    break;
+                default:
+                    throw new ScannerError('Invalid escape sequence', this.position.line, this.position.column);
+            }
+
+            pos++;
+            start = pos;
+        }
+
+        result += text.substring(start, pos);
+        return result;
+    }
+
+    scanIdentifier() {
+        this.text = this.scanUntil((ch) => !isIdentifierPart(ch));
+        return this.kind = keywords.get(this.text) ?? Kind.Identifier;
+    }
+
+    scanVariable() {
+        this.text = '$';
+        this.text += this.scanUntil((ch) => !isIdentifierPart(ch));
+        return this.kind = Kind.Variable;
+    }
+
+    /**
+   * Returns the zero-based line/column from the given offset
+   * (binary search thru the token start locations)
+   * @param offset the character position in the document
+   */
+    positionFromOffset(offset: number): Position {
+        let position = { line: 0, column: 0, offset: 0 };
+
+        // eslint-disable-next-line keyword-spacing
+        if (offset < 0 || offset > this.#length) {
+            return { line: position.line, column: position.column };
+        }
+
+        let first = 0;    // left endpoint

```

Comment:
One line has 4 spaces before `//` and the next line has 3 spaces before `//` so it seems random. Can there just be 1 space? I don't see a readability benefit in adding more spaces in this case.

Reply:
reducing whitespace to 1